src/OrleansAzureUtils/MultiClusterNetwork/AzureTableBasedGossipChannel.cs

```diff
@@ -0,0 +1,202 @@
+        {
+            logger.Verbose("-PushAndPull pushed:{0}", pushed);
+
+            try
+            {
+                var entriesFromStorage = await tableManager.FindAllGossipTableEntries();
+                var delta = await DiffAndWriteBack(pushed, entriesFromStorage);
+
+                logger.Verbose("-PushAndPull pulled delta:{0}", delta);
+
+                return delta;
+            }
+            catch (Exception e)
+            {
+                logger.Info("-PushAndPull encountered exception {0}", e);
+
+                throw e;
+            }
+        }
+
+        internal async Task<MultiClusterData> DiffAndWriteBack(MultiClusterData dataToPush, IEnumerable<GossipTableEntry> entriesFromStorage)
+        {
+            GossipTableEntry configRow = null;
+            var gateways = new Dictionary<SiloAddress, KeyValuePair<GatewayEntry, GossipTableEntry>>();
+            MultiClusterConfiguration returnedConfiguration = null;
+
+            // collect left-hand side data
+            MultiClusterConfiguration configToPush = dataToPush.Configuration;
+            foreach (var gatewayEntry in dataToPush.Gateways.Values)
+            {
+                if (!gatewayEntry.Expired)
+                    gateways[gatewayEntry.SiloAddress] = new KeyValuePair<GatewayEntry, GossipTableEntry>(gatewayEntry, null);
+            }
+
+            foreach (var tableEntry in entriesFromStorage)
+            {
+                if (tableEntry.RowKey.Equals(GossipTableEntry.CONFIGURATION_ROW))
+                {
+                    configRow = tableEntry;
+                    // interpret empty admin timestamp by taking the azure table timestamp instead
+                    // this allows an admin to inject a configuration by editing table more easily
+                    if (configRow.GossipTimestamp == default(DateTime))
+                        configRow.GossipTimestamp = configRow.Timestamp.UtcDateTime;
+                }
+                else
+                {
+                    try
+                    {
+                        tableEntry.UnpackRowKey();
+                        UpdateDictionaryRightValue(gateways, tableEntry.SiloAddress, tableEntry);
+                    }
+                    catch (Exception exc)
+                    {
+                        logger.Error(
+                            ErrorCode.AzureTable_61,
+                            string.Format("Intermediate error parsing GossipTableEntry: {0}. Ignoring this entry.", tableEntry),
+                            exc);
+                    }
+                }
+            }
+
+            var writeback = new List<Task>();
+            var sendback = new Dictionary<SiloAddress,GatewayEntry>();
+
+            if (configToPush != null &&
+                (configRow == null || configRow.GossipTimestamp < configToPush.AdminTimestamp))
+            {
+                // push configuration
+                if (configRow == null)
```

Comment:
Upsert won't work for two reasons: (a) information goes both ways, not just to storage, and (2) the update to storage is conditional on the timestamp being newer.